### PR TITLE
ConverterStrategyHandler

### LINKED ISSUE #22: ConverterStrategyHandler
In some situations you will need a converter which is choosing the correct converter by some circumstances (e.g. correct object type, country, you are converting for, language based or other environment settings).

So, there should be a `ConverterStrategyHandler`, which is of course a Converter itself but uses other converters to do the work.
Maybe the strategy which converter should be chosen could be something like this:
```php
interface DetermineConversionService 
{
    /**
     * @returns string which is the associated array key of the list of converters
     */
    public function determine(object $source, object $ctx): string
}
```

Implementation of `ConverterStrategyHandler` could be something like this:
```php
class ConverterStrategyHandler implements Converter
{
    public function __construct(
        private DetermineConversionService $determineConversionService,
        private array $converters = []
    ) {
    }

    public function convert(object $source, object $ctx): object
    {
        if ($criterium = $this->determineConversionService->determine($source, $ctx)) {
            return $this->converters[$criterium]->convert($source, $ctx);
        }

        return null;
    }
}
```

### PR DESCRIPTION
**Pull request by mike4git**
A first approach for discussion and think about better solutions.

### CODE REVIEW

File: src/Converter/ConverterStrategyHandler.php [hunk 1]
```diff
@@ -0,0 +1,42 @@
...
+/**
+ * @template S of object
+ * @template T of object
+ * @template C of object
+ * @implements Converter<S, T, C>
+ */
+class ConverterStrategyHandler implements Converter
+{
+    /**
+     * @param array<Converter<S, T, C>> $converters
```
Comment (jdreesen) [inline src/Converter/ConverterStrategyHandler.php]:
```suggestion
     * @param array<string, Converter<S, T, C>> $converters
```

File: tests/Converter/ConverterStrategyHandlerTest.php [hunk 2]
```diff
@@ -0,0 +1,78 @@
...
+        $person = new Person();
+
+        $this->determineConverterService->determine($user, null)->willReturn('testKey');
+        $this->converter->convert($user, null)->willReturn($person)->shouldBeCalled();
+
+        $this->strategyHandler->convert($user);
+
+    }
+
+    public function testConvert_excetional_case(): void
```
Comment (jdreesen) [inline tests/Converter/ConverterStrategyHandlerTest.php]:
```suggestion
    public function testConvert_exceptional_case(): void
```

File: src/Strategy/DetermineConverterService.php [hunk 3]
```diff
@@ -0,0 +1,14 @@
...
+namespace Neusta\ConverterBundle\Strategy;
+
+/**
+ * @template S of object
+ * @template C of object
+ */
+interface DetermineConverterService
+{
+    public function determine(object $source, ?object $ctx = null): string;
+}
```
Comment (jdreesen) [inline src/Strategy/DetermineConverterService.php]: What do you think of:

```suggestion
interface ConverterSelector
{
    public function selectConverter(object $source, ?object $ctx = null): string;
}
```

**Comment by jdreesen** [comment]
Please rebase to get rid of all those unrelated changes.

**Comment by jdreesen** [comment]
I'm going to rebase the branch to fix it up now. Please have a look afterwards if everything is as you intended.

**Review by jdreesen — DISMISSED** [review]
I think this is fine for now.

**Comment by mike4git** [comment]
So, hopefully now only necessary changes are inside the branch. Please review it now - thx.

**Review by jdreesen — APPROVED** [review]

**Pull request merged**